source/opt/scalar_replacement_pass.cpp

```diff
@@ -172,10 +172,14 @@ bool ScalarReplacementPass::ReplaceWholeDebugDeclare(
   // Add DebugValue instruction with Indexes operand and Deref operation.
   int32_t idx = 0;
   for (const auto* var : replacements) {
+    Instruction* insert_before = var->NextNode();
+    while (insert_before->opcode() == SpvOpVariable)
+      insert_before = insert_before->NextNode();
+    assert(insert_before != nullptr && "unexpected end of list");
```

Comment:
Just out of curiosity.
Is there any reason to add `DebugValue` after all variables?
As far as I remember, the instruction layout allows us to put `DebugValue` between `OpVariable`s

Reply:
Because these are "nonsemantic" instructions, they are more restricted than the OpenCL.DebugInfo.100 instructions:

From https://github.com/KhronosGroup/SPIRV-Registry/blob/master/nonsemantic/NonSemantic.Shader.DebugInfo.100.asciidoc#binary-form 

DebugScope, DebugNoScope, DebugDeclare, DebugValue, DebugLine, DebugNoLine, and DebugFunctionDefinition instructions can interleave with the instructions within a function, but must appear within valid locations in a block as required by SPV_KHR_non_semantic_info. In particular this means **they cannot come before any OpPhi or function-level variable declarations in a block**, and they cannot come after a Merge Instruction.